src/Dialect/ONNX/ElementsAttr/DisposablePool.cpp

```diff
@@ -114,17 +114,24 @@ void DisposablePool::scrub(ModuleOp moduleOp, OpAttrDictionary opsAttrs) {
       return llvm::make_range(batchBegin, batchEnd);
     };
     // Parallel worker body: Fetch and process batches until there are no more.
-    auto work = [&fetchBatch](size_t threadNumber) {
+    auto work = [&fetchBatch, &translationMutex](size_t threadNumber) {
       for (;;) {
         auto batch = fetchBatch();
         if (batch.empty())
           break;
         for (auto &[id, translation] : batch) {
           auto &[disposable, dense] = translation;
           dense = disposable.toDenseElementsAttr();
-          // TODO: Consider calling disposable.dispose() here to free up memory
-          //       on the go to make memory available to create the next
-          //       DenseElementsAttr. In that case should we lock mutex?
+        }
+
+        // Delete DisposableElementsAttr for each batch with lock after
+        // conversion.
+        {
+          const std::lock_guard<std::mutex> lock(translationMutex);
+          for (auto &[id, translation] : batch) {
```

Comment:
Can we lock  the original `for-loop` at Line 122 instead of creating a new loop? I guess calling `dispose()` inside the original loop can help reduce the peak memory.

Reply:
Yes. I used the lock for each batch because I was concerned frequent lock might affect the compile time and I confirmed the spike can be removed with the lock for each batch.
However, from my test, it seems to be no performance difference with original loop. I will update.